workitem/workitem_repository.go

```diff
@@ -83,6 +103,116 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
 	return nil
 }
 
+// Reorder reorders the given work item in storage. Version must be the same as the one int the stored version
+// returns NotFoundError, VersionConflictError, ConversionError or InternalError
+func (r *GormWorkItemRepository) Reorder(ctx context.Context, before *string, wi app.WorkItem) (*app.WorkItem, error) {
+	var order float64
+	res := WorkItem{}
+	beforeItem := WorkItem{}
+	afterItem := WorkItem{}
+	id, err := strconv.ParseUint(wi.ID, 10, 64)
+	if err != nil || id == 0 {
+		return nil, errors.NewNotFoundError("work item", wi.ID)
+	}
+
+	log.Printf("looking for id %d", id)
+	tx := r.db.First(&res, id)
+	if tx.RecordNotFound() {
+		log.Printf("not found, res=%v", res)
+		return nil, errors.NewNotFoundError("work item", wi.ID)
+	}
+	if tx.Error != nil {
+		return nil, errors.NewInternalError(err.Error())
+	}
+	if res.Version != wi.Version {
+		return nil, errors.NewVersionConflictError("version conflict")
+	}
+
+	wiType, err := r.wir.LoadTypeFromDB(wi.Type)
+	if err != nil {
+		return nil, errors.NewBadParameterError("Type", wi.Type)
+	}
+
+	if before != nil {
+		beforeId, err := strconv.ParseUint(*before, 10, 64)
+		if err != nil || beforeId == 0 {
+			return nil, errors.NewNotFoundError("work item", *before)
+		}
+		log.Printf("looking for id %d", beforeId)
+		tx = r.db.First(&beforeItem, beforeId)
+		if tx.RecordNotFound() {
+			log.Printf("not found, res=%v", beforeItem)
+			return nil, errors.NewNotFoundError("work item", string(beforeId))
+		}
+		if tx.Error != nil {
+			return nil, errors.NewInternalError(err.Error())
+		}
+		beforeOrder, err := strconv.ParseFloat(fmt.Sprintf("%v", beforeItem.Fields[SystemOrder]), 64)
+		if err != nil {
+			return nil, errors.NewBadParameterError("data.attributes.order", res.Fields[SystemOrder])
+		}
+		tx2 := r.db.Where("fields -> 'order' < ?", beforeItem.Fields[SystemOrder]).Order("fields->'order' desc", true).Last(&afterItem)
+		if afterItem.ID == 0 {
+			// The item is moved to first position
+			order = (0 + beforeOrder) / 2
+		} else {
+			if tx2.RecordNotFound() {
+				log.Printf("not found, res=%v", beforeItem)
+				return nil, errors.NewNotFoundError("work item", *before)
+			}
+			if tx2.Error != nil {
+				return nil, errors.NewInternalError(err.Error())
+			}
+		}
+		afterOrder, err := strconv.ParseFloat(fmt.Sprintf("%v", afterItem.Fields[SystemOrder]), 64)
+		if err != nil {
+			return nil, errors.NewBadParameterError("data.attributes.order", res.Fields[SystemOrder])
+		}
+		order = (beforeOrder + afterOrder) / 2
+	} else {
+		// the item is moved at last position
+
```

Comment:
This logic still executes even if WI is already at last. I understand there is no such case via UI.